src/Uno.UI/UI/Xaml/Controls/Image/ImageSizeHelper.cs

```diff
@@ -1,122 +1,159 @@
+						: finalSize.Width / child.Width; // child is taller than parent
 					child.Width *= uniformToFillScale;
 					child.Height *= uniformToFillScale;
 					break;
 
 				case Uniform:
-					var uniformScale = (child.Width * parent.Height > child.Height * parent.Width)
-						? parent.Width / child.Width // child is taller than parent
-						: parent.Height / child.Height; // child is flatter than parent
+					var uniformScale = (child.Width * finalSize.Height > child.Height * finalSize.Width)
+						? finalSize.Width / child.Width // child is taller than parent
+						: finalSize.Height / child.Height; // child is flatter than parent
 					child.Width *= uniformScale;
 					child.Height *= uniformScale;
 					break;
 
 				case Fill:
-					child.Width = parent.Width;
-					child.Height = parent.Height;
+					child.Width = finalSize.Width;
+					child.Height = finalSize.Height;
 					break;
 
 				case None:
 					break;
 			}
 		}
 
-		public static void ArrangeSource(this Image image, Windows.Foundation.Rect parent, ref Windows.Foundation.Rect child)
+		public static void ArrangeSource(this Image image, Size finalSize, ref Rect child)
 		{
+			var stretch = image.Stretch;
 			// In order to match UWP behaviors, in some specific cases the image must be left align
-			var isForcedLeft = (image.Stretch == None || image.Stretch == UniformToFill) && parent.Width <= child.Width;
+			//var isForcedLeft = (stretch == None || stretch == UniformToFill) && finalSize.Width <= child.Width;
+			var isForcedLeft = false;
+
+			double GetPositionForDimension(bool isForcedToZero, double imageDimension, double calculatedPosition)
+			{
+				if (isForcedToZero && !double.IsNaN(imageDimension))
+				{
+					return 0;
+				}
+
+				return calculatedPosition;
+			}
 
 			switch (image.HorizontalAlignment)
 			{
 				case HorizontalAlignment.Left:
 					child.X = 0;
 					break;
 				case HorizontalAlignment.Right:
-					child.X = isForcedLeft && !Double.IsNaN(image.Width) ? 0 : parent.Width - child.Width;
+					child.X = GetPositionForDimension(isForcedLeft, image.Width, finalSize.Width - child.Width);
 					break;
 				case HorizontalAlignment.Center:
-					child.X = isForcedLeft && !Double.IsNaN(image.Width) ? 0 : (parent.Width * 0.5f) - (child.Width * 0.5f);
+					child.X = GetPositionForDimension(isForcedLeft, image.Width, (finalSize.Width - child.Width) * 0.5f);
 					break;
 				case HorizontalAlignment.Stretch:
-					child.X = isForcedLeft ? 0 : (parent.Width * 0.5f) - (child.Width * 0.5f);
+					child.X = GetPositionForDimension(isForcedLeft, -1, (finalSize.Width - child.Width) * 0.5f);
 					break;
 			}
 
 			// In order to match UWP behaviors, in some specific cases the image must be top align
-			var isForcedTop = (image.Stretch == None || image.Stretch == UniformToFill) && parent.Height <= child.Height;
+			//var isForcedTop = (stretch == None || stretch == UniformToFill) && finalSize.Height <= child.Height;
+			var isForcedTop = false;
```

Comment:
idem https://github.com/unoplatform/uno/pull/2123/files#r349345436

Reply:
The logic has been simplified.